### PR TITLE
kube-apiserver identity

### PR DESCRIPTION
**Pull request by roycaihw**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Implements this [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/1965-kube-apiserver-identity/README.md). 

**Does this PR introduce a user-facing change?**:

```release-note
kube-apiserver now maintains a Lease object to identify itself: 
- The feature is under feature gate `APIServerIdentity`. 
- Two flags are added to kube-apiserver: `lease-duration-seconds`, `lease-renew-interval-seconds`
A new controller `lease-garbage-collector` is added to kube-controller-manager, which deleted expired kube-apiserver Lease objects:
- The feature is under feature gate `APIServerIdentity`.
- A flag is added to kube-controller-manager: `lease-resync-period`
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/1965-kube-apiserver-identity/README.md
```

/sig api-machinery

### CODE REVIEW

File: staging/src/k8s.io/client-go/pkg/controllers/lease/controller.go [hunk 1]
```diff
@@ -170,10 +182,14 @@ func (c *controller) ensureLease() (*coordinationv1.Lease, bool, error) {
...
+		leaseToUpdate, blockUpdate, err := c.newLease(base)
+		if !blockUpdate {
+			var lease *coordinationv1.Lease
+			lease, err = c.leaseClient.Update(context.TODO(), leaseToUpdate, metav1.UpdateOptions{})
+			if err == nil {
+				c.latestLease = lease
+				return nil
+			}
 		}
 		klog.Errorf("failed to update node lease, error: %v", err)
```
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/pkg/controllers/lease/controller.go:194]: Not "node lease"

File: staging/src/k8s.io/client-go/pkg/controllers/lease/controller.go [hunk 2]
```diff
@@ -170,10 +182,14 @@ func (c *controller) ensureLease() (*coordinationv1.Lease, bool, error) {
 // call this once you're sure the lease has been created
 func (c *controller) retryUpdateLease(base *coordinationv1.Lease) error {
 	for i := 0; i < maxUpdateRetries; i++ {
-		lease, err := c.leaseClient.Update(context.TODO(), c.newLease(base), metav1.UpdateOptions{})
-		if err == nil {
-			c.latestLease = lease
-			return nil
+		leaseToUpdate, blockUpdate, err := c.newLease(base)
+		if !blockUpdate {
```
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/pkg/controllers/lease/controller.go:186]: Can you use the opposite condition? i.e.,

```golang
if blockUpdate {
  klog.Error(...)
  continue
}
lease, err = c.leaseClient.Update(context.TODO(), leaseToUpdate, metav1.UpdateOptions{})
if err == nil {
  c.latestLease = lease
  return nil
}
klog.Error(...)
```


Also move the onRepeatedHeartbeatFailure check to the beginning of the for loop to keep the semantics.

**Review by caesarxuchao — COMMENTED** [review]
Can you split this PR into smaller pieces:
* moving the lease controller, and update the kubelet code
* add apiserver lease controller
* add apiserver lease garbage collector

**Comment by roycaihw** [comment]
> Can you split this PR into smaller pieces:

>moving the lease controller, and update the kubelet code
add apiserver lease controller
add apiserver lease garbage collector

That sounds much better. Will do

**Comment by roycaihw** [comment]
Closing in favor of https://github.com/kubernetes/kubernetes/pull/95428

File: pkg/controller/leasegarbagecollector/leasegarbagecollector.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with: %v", key, err))
+	c.queue.AddRateLimited(key)
+	return true
+}
+
+func (c *Controller) sync(name string) error {
+	lease, err := c.leaseLister.Leases(corev1.NamespaceAPIServerLease).Get(name)
```
Comment (roycaihw) [inline pkg/controller/leasegarbagecollector/leasegarbagecollector.go:116]: use kubeclient to get latest object from apiserver

**Pull request closed without merging**